lib/ansible/plugins/action/validate_arg_spec.py

```diff
@@ -0,0 +1,153 @@
+
+    def check_for_errors(self):
+        if not self.arg_validation_errors:
+            return
+
+        msg = 'Validation of arguments failed:\n%s' % '\n'.join(self.arg_validation_errors)
+        raise AnsibleArgSpecError(msg, argument_errors=self.arg_validation_errors)
+
+
+class ActionModule(ActionBase):
+    ''' Validate a arg spec'''
+
+    TRANSFERS_FILES = False
+
+    # WARNING: modifies argument_spec
+    def build_args(self, argument_spec, task_vars):
+        args = {}
+        for key, attrs in iteritems(argument_spec):
+            if attrs is None:
+                argument_spec[key] = {'type': 'str'}
+
+            if key in task_vars:
+                if isinstance(task_vars[key], string_types):
+                    value = self._templar.do_template(task_vars[key])
+                    if value:
+                        args[key] = value
+                else:
+                    args[key] = task_vars[key]
+            elif attrs:
+                if 'aliases' in attrs:
+                    for item in attrs['aliases']:
+                        if item in task_vars:
+                            args[key] = self._templar.do_template(task_vars[key])
+                elif 'default' in attrs and key not in args:
+                    args[key] = attrs['default']
+        return args
+
+    def run(self, tmp=None, task_vars=None):
+        '''
+        Validate an arg spec
+
+        :arg tmp: Deprecated. Do not use.
+        :arg dict task_vars: A dict of task variables.
+            Valid args include 'argument_spec', 'provided_arguments'
+        :return: An action result dict, including a 'argument_errors' key with a
+            list of validation errors found.
+        '''
+        if task_vars is None:
+            task_vars = dict()
+
+        result = super(ActionModule, self).run(tmp, task_vars)
+        del tmp  # tmp no longer has any effect
+
+        # This action can be called from anywhere, so pass in some info about what it is
+        # validating args for so the error results make some sense
+        result['validate_args_context'] = self._task.args.get('validate_args_context', {})
+
+        if 'argument_spec' not in self._task.args:
+            raise AnsibleError('"argument_spec" arg is required in args: %s' % self._task.args)
+
+        # get the task var called argument_spec
+        argument_spec_data = self._task.args.get('argument_spec')
+
+        # include arg spec data dict in return, use a copy since we 'pop' from it
+        # and modify it in place later.
+        orig_argument_spec_data = argument_spec_data.copy()
+
+        # then get the 'argument_spec' item from the dict in the argument_spec task var
+        # everything left in argument_spec_data is modifiers
+        argument_spec = argument_spec_data.pop('argument_spec', [])
+
+        # the values that were passed in and will be checked against argument_spec
+        provided_arguments = self._task.args.get('provided_arguments', {})
```

Comment:
this should just use task_vars as commented above, no need to pass a specific argument

Reply:
provided_args allows a task to pass in an explicit data structure to validate without it needing to exist somewhere in ansible vars.

Reply:
'normal use case' is validating 'current vars', which user cannot really pass in, they can only acces subsets of 'task_vars' ... may be ok as option (what use case?) but it should at least default to task_vars .. which is what we need to validate in the case from the proposal